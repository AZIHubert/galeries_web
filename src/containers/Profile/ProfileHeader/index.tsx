import * as React from 'react';

import Information from './Information';
import ProfilePicture from './ProfilePicture';

import { Container } from './styles';

const ProfileHeader = () => (
  <Container>
    <ProfilePicture />
    <Information />
  </Container>
);

export default ProfileHeader;
