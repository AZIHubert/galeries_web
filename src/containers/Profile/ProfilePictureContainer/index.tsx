import * as React from 'react';

const ProfilPictureContainer = () => (
  <div>

  </div>
);

export default ProfilPictureContainer;
