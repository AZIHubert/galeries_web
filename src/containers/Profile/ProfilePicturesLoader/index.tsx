import * as React from 'react';
import { useSelector } from 'react-redux';
import DotLoader from 'react-spinners/DotLoader';
import { CSSTransition } from 'react-transition-group';

import theme from '#helpers/theme';

import { profilePicturesStatusSelector } from '#store/selectors';

import {
  Container,
  Fader,
} from './styled';

const ProfilePicturesLoader = () => {
  const profilePicturesStatus = useSelector(profilePicturesStatusSelector);
  const isFetching = React.useMemo(() => profilePicturesStatus === 'fetching', [profilePicturesStatus]);

  return (
    <CSSTransition
      classNames='fade'
      in={isFetching}
      timeout={300}
      unmountOnExit
    >
      <Fader>
        <Container>
          <DotLoader
            color={theme.colors.primary}
            size={50}
          />
        </Container>
      </Fader>
    </CSSTransition>
  );
};

export default ProfilePicturesLoader;
