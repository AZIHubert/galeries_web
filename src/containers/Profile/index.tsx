import * as React from 'react';

import { UserContext } from '#contexts/UserContext';

import {
  ImageI,
  ProfilePictureI,
  UserI,
} from '#helpers/interfaces';

import ProfilePictureContainer from './ProfilePictureContainer';

const profilePicture
: (user: UserI | null) => any = (user: UserI | null) => {
  if (user && user.currentProfilePicture) {
    return user.currentProfilePicture.cropedImage.signedUrl;
  }
  if (user && user.defaultProfilePicture) {
    return user.defaultProfilePicture;
  }
  return '#ressources/images/defaultProfilePicture.png';
};

const Profile = () => {
  const [selectedFile, setSelectedFile] = React.useState<File | null>(null);
  const fileInputRef = React.useRef<HTMLInputElement | null>(null);
  const { setUser, user } = React.useContext(UserContext);
  React.useEffect(() => {
    if (selectedFile) {
      const newImage: ImageI = {
        bucketName: 'bucketName',
        fileName: selectedFile.name,
        format: 'jpg',
        height: 2340,
        id: '2',
        signedUrl: 'http://newprofilepicture.com/',
        size: 100857,
        width: 1080,
      };
      const newProfilePicture: ProfilePictureI = {
        createdAt: new Date(),
        cropedImage: newImage,
        id: '2',
        originalImage: newImage,
        pendingImage: newImage,
      };
      setUser((prevState) => {
        if (prevState) {
          return {
            ...prevState,
            currentProfilePictureId: newProfilePicture.id,
            currentProfilePicture: newProfilePicture,
            profilePictures: [
              newProfilePicture,
              ...prevState.profilePictures,
            ],
          };
        }
        return null;
      });
    }
  }, [selectedFile]);
  const addFile = (e: React.ChangeEvent<HTMLInputElement>) => {
    if (e.target.files) {
      setSelectedFile(e.target.files[0]);
    }
  };
  const inputClick = () => {
    if (fileInputRef.current) { fileInputRef.current.click(); }
  };
  return (
    <div>
      <img
        data-testid='currentProfilePicture'
        src={profilePicture(user)}
        alt='current profile picture'
      />
      <p
        data-testid='userNameText'
      >
        {user ? user.userName : 'user name'}
      </p>
      <input
        accept="image/*"
        data-testid='inputFile'
        onChange={addFile}
        ref={fileInputRef}
        type="file"
      />
      <button
        data-testid="addProfilePictureButton"
        onClick={inputClick}
      >
        Add a profile picture
      </button>
      <button>
        Edit your info
      </button>
      <ProfilePictureContainer />
    </div>
  );
};

export default Profile;
