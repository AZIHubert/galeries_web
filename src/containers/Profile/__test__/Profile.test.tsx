import {
  cleanup,
  fireEvent,
  render,
  screen,
} from '@testing-library/react';
import * as React from 'react';
import renderer from 'react-test-renderer';

import {
  UserContext,
  UserProvider,
} from '#contexts/UserContext';

import {
  ImageI,
  ProfilePictureI,
  UserI,
} from '#helpers/interfaces';

import Profile from '../index';

const newImage: ImageI = {
  bucketName: 'bucketName',
  fileName: 'fileName',
  format: 'jpg',
  height: 2340,
  id: '1',
  signedUrl: 'http://url.com',
  size: 100857,
  width: 1080,
};
const newProfilePicture: ProfilePictureI = {
  createdAt: new Date('2021-02-05 18:42:02.528+01'),
  cropedImage: newImage,
  id: '1',
  originalImage: newImage,
  pendingImage: newImage,
};

const newUser: UserI = {
  createdAt: new Date('2021-02-05 18:42:02.528+01'),
  currentProfilePicture: null,
  currentProfilePictureId: null,
  defaultProfilePicture: null,
  email: 'user@email.com',
  facebookId: null,
  googleId: null,
  id: '1',
  profilePictures: [newProfilePicture],
  role: 'admin',
  updatedAt: null,
  userName: 'user',
};

const Container = ({
  defaultUser = newUser,
}: {
  defaultUser?: UserI
}) => {
  const { setUser } = React.useContext(UserContext);
  React.useEffect(() => {
    setUser(defaultUser);
  }, []);
  return (
    <Profile />
  );
};

describe('Profile', () => {
  afterEach(cleanup);
  it('renders without crashing', () => {
    const tree = renderer.create(<Profile />).toJSON();
    expect(tree).toMatchSnapshot();
  });
  it('should render local profile picture if currentProfilePictureId and defaultProfilePicture are null', () => {
    const { getByTestId } = render(
      <UserProvider>
        <Container />
      </UserProvider>,
    );
    const currentProfilePicture = getByTestId('currentProfilePicture');
    expect(currentProfilePicture).toHaveAttribute('src', '#ressources/images/defaultProfilePicture.png');
  });
  it('should render default profile picture if currentProfilePictureId is null and defaultProfilePicture is set', () => {
    const userWithDefaultProfilePicture: UserI = {
      ...newUser,
      defaultProfilePicture: 'defaultProfilePicture',
    };
    const { getByTestId } = render(
      <UserProvider>
        <Container
          defaultUser={userWithDefaultProfilePicture}
        />
      </UserProvider>,
    );
    const currentProfilePicture = getByTestId('currentProfilePicture');
    expect(currentProfilePicture).toHaveAttribute('src', 'defaultProfilePicture');
  });
  it('should render profile picture if currentProfilePicture is set', () => {
    const userWithCurrentProfilePicture: UserI = {
      ...newUser,
      currentProfilePicture: newProfilePicture,
      currentProfilePictureId: newProfilePicture.id,
    };
    const { getByTestId } = render(
      <UserProvider>
        <Container
          defaultUser={userWithCurrentProfilePicture}
        />
      </UserProvider>,
    );
    const currentProfilePicture = getByTestId('currentProfilePicture');
    expect(currentProfilePicture).toHaveAttribute('src', newProfilePicture.originalImage.signedUrl);
  });
  it('should render profile picture, if currentProfilePicture and defaultProfilePicture are set', () => {
    const userWithCurrentProfilePicture: UserI = {
      ...newUser,
      currentProfilePicture: newProfilePicture,
      currentProfilePictureId: newProfilePicture.id,
      defaultProfilePicture: 'defaultProfilePicture',
    };
    const { getByTestId } = render(
      <UserProvider>
        <Container
          defaultUser={userWithCurrentProfilePicture}
        />
      </UserProvider>,
    );
    const currentProfilePicture = getByTestId('currentProfilePicture');
    expect(currentProfilePicture).toHaveAttribute('src', newProfilePicture.originalImage.signedUrl);
  });
  it('should render username', () => {
    const { getByTestId } = render(
      <UserProvider>
        <Container />
      </UserProvider>,
    );
    const userNameText = getByTestId('userNameText');
    expect(userNameText).toHaveTextContent(newUser.userName);
  });
  it('should set picture if currentProfileProfile is null', () => {
    const { getByTestId } = render(
      <UserProvider>
        <Container />
      </UserProvider>,
    );
    const profilePictureButton = getByTestId('profilePictureButton');
    fireEvent.click(profilePictureButton);
    const currentProfilePicture = screen.getByTestId('currentProfilePicture');
    expect(currentProfilePicture).toHaveAttribute('src', newProfilePicture.originalImage.signedUrl);
  });
  it('should remove profile if currentProfilePicture is not null', () => {
    const userWithCurrentProfilePicture: UserI = {
      ...newUser,
      currentProfilePicture: newProfilePicture,
      currentProfilePictureId: newProfilePicture.id,
    };
    const { getByTestId } = render(
      <UserProvider>
        <Container
          defaultUser={userWithCurrentProfilePicture}
        />
      </UserProvider>,
    );
    const profilePictureButton = getByTestId('profilePictureButton');
    fireEvent.click(profilePictureButton);
    const currentProfilePicture = screen.getByTestId('currentProfilePicture');
    expect(currentProfilePicture).toHaveAttribute('src', '#ressources/images/defaultProfilePicture.png');
  });
  it('should add a profile picture', () => {
    const file = new File(['blob'], 'image.png', { type: 'image/png' });
    const { getByTestId, getAllByTestId } = render(
      <UserProvider>
        <Container />
      </UserProvider>,
    );
    const previousProfilePictures = getAllByTestId('profilePicture');
    const inputFile = getByTestId('inputFile');
    fireEvent.change(inputFile, { target: { files: [file] } });
    const profilePictures = screen.getAllByTestId('profilePicture');
    const currentProfilePicture = screen.getByTestId('currentProfilePicture');
    expect(profilePictures.length).toBe(previousProfilePictures.length + 1);
    expect(currentProfilePicture).toHaveProperty('src', 'http://newprofilepicture.com/');
  });
});
