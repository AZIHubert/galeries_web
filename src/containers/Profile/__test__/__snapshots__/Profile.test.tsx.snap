// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Profile renders without crashing 1`] = `
<div>
  <img
    alt="current profile picture"
    data-testid="currentProfilePicture"
    src="#ressources/images/defaultProfilePicture.png"
  />
  <p
    data-testid="userNameText"
  >
    user name
  </p>
  <input
    accept="image/*"
    data-testid="inputFile"
    onChange={[Function]}
    type="file"
  />
  <button
    data-testid="addProfilePictureButton"
    onClick={[Function]}
  >
    Add a profile picture
  </button>
  <button>
    Edit your info
  </button>
</div>
`;
